### PR TITLE
Fix 51 issue

### PR DESCRIPTION
**Pull request by IlhamKhabibullin**
I think there is still a lot of work I have to do, now i want to get sure that i moving in right direction
@ffMathy you the one who I expect will assure me

### CODE REVIEW

**Comment by IlhamKhabibullin** [comment]
@ffMathy I am planning to continue working on this PR after having your feedback

**Comment by ffMathy** [comment]
Direction seems OK, if you can get the existing breaking tests passing.

Thank you so much for looking into this! <3

**Comment by IlhamKhabibullin** [comment]
May be we should replace `received` method with some Symbol? That would free us from necessity of `disabledFor` stuff? 

I will try without it, but just think about that. 
Sequlize library uses this approach for operators in db queries, operators like $where.

**Comment by IlhamKhabibullin** [comment]
@ffMathy I have made required fixes

"class method received" test was broken because I was throwing error in `set` of `receivedProxy`, now it does same thing as simple `proxy` of `context` 

Two tests for disableFor also was broken, it actually seems like they was not working correctly, get, set and apply of disableProxy was sending its `arguments` object into get, set and apply of `Context`, which ones was expect not same list of arguments, for example, `context.get` was expecting only `property`, but it was getting `target, this, property`, and сonsequently getting `target` where it was expecting `property`

I fixed that

also it seems like `received` called with `disableFor` was never recording as `getPropertyState` inside `InitialState`, not sure if it was broken always or I did that, but anyway I fixed that as well

waiting for your feedback 

[

](url)

**Comment by ffMathy** [comment]
So all tests pass?

**Comment by ffMathy** [comment]
If so, that's impressive!

File: src/states/SetPropertyState.ts [hunk 1]
```diff
@@ -54,6 +54,6 @@ export class SetPropertyState implements ContextState {
     }
 
     get(context: Context, property: PropertyKey): undefined {
-        return void 0;
+        return void 0; // TODO:  probably better throw error since this should never happen
```
Comment (ffMathy) [inline src/states/SetPropertyState.ts]: In that case, feel free to throw here

Comment (IlhamKhabibullin) [reply src/states/SetPropertyState.ts]: done

**Comment by ffMathy** [comment]
Do all tests pass?

**Comment by ffMathy** [comment]
Or is it just the tests for issue #51?

**Comment by IlhamKhabibullin** [comment]
All tests pass, 39 exact number, 1 tests skipped, but not by me

**Comment by ffMathy** [comment]
@IlhamKhabibullin the build fails on the build server.

> spec/issues/59.test.ts(16,31): error TS2554: Expected 0 arguments, but got 1.

**Comment by IlhamKhabibullin** [comment]
researching on it

ср, 13 нояб. 2019 г. в 14:18, Mathias Lykkegaard Lorenzen <
notifications@github.com>:

**Comment by IlhamKhabibullin** [comment]
isn't it because of previous PR
https://github.com/ffMathy/FluffySpoon.JavaScript.Testing.Faking/commit/92f7e60b871714f60ace5bb3ddc7ce6d5e0908ac

ср, 13 нояб. 2019 г. в 14:22, Ильхам Хабибуллин <mailbox.ilham@gmail.com>:

**Comment by IlhamKhabibullin** [comment]
author says that change adds an example of failing received with optional
arguments, but it looks more like imperfections of typescript

ср, 13 нояб. 2019 г. в 14:41, Ильхам Хабибуллин <mailbox.ilham@gmail.com>:

**Comment by ffMathy** [comment]
But didn't you mention that the tests were passing? How did you get it running on your machine?

**Comment by IlhamKhabibullin** [comment]
yeah, that code was added 8 days ago, and I did my fork earlier, so on my
fork's master I wasn't having this broken change

ср, 13 нояб. 2019 г. в 14:55, Mathias Lykkegaard Lorenzen <
notifications@github.com>:

**Comment by ffMathy** [comment]
Ah that makes sense then. Do you have the courage to try to fix the optional issue too? 🙇

**Pull request merged**